src/ucp/proto/proto_multi.c

```diff
@@ -151,6 +151,69 @@ ucp_proto_multi_init_flush_sys_dev_mask(const ucp_rkey_config_key_t *key)
     return UCS_BIT(key->sys_dev & ~UCP_SYS_DEVICE_FLUSH_BIT);
 }
 
+static ucp_lane_index_t ucp_proto_multi_filter_net_devices(
+        ucp_lane_index_t num_lanes, const ucp_proto_init_params_t *params,
+        const ucp_proto_common_tl_perf_t *tl_perfs, int fixed_first_lane,
+        ucp_lane_index_t *lanes, ucp_proto_perf_node_t **perf_nodes)
+{
+    double max_bandwidth;
+    ucp_lane_index_t i, lane, num_identical_devs, seed, num_filtered_lanes;
+    ucp_lane_map_t lane_map;
+    ucs_sys_device_t sys_dev;
+    ucs_sys_device_t sys_devs[UCP_PROTO_MAX_LANES];
+    const uct_tl_resource_desc_t *tl_rsc;
+
+    for (lane_map = 0, max_bandwidth = 0, i = 0; i < num_lanes; ++i) {
+        lane = lanes[i];
+        if (!ucp_proto_common_is_net_dev(params, lane)) {
+            continue;
+        }
+
+        lane_map     |= UCS_BIT(lane);
+        max_bandwidth = ucs_max(max_bandwidth, tl_perfs[lane].bandwidth);
+    }
+
+    num_identical_devs = 0;
```

Comment:
it is a bit confusing name, maybe just num_net_devs? Because it represents number of different devices, not identical